service/collector/cluster_transition.go

```diff
@@ -0,0 +1,182 @@
+package collector
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	infrastructurev1alpha2 "github.com/giantswarm/apiextensions/pkg/apis/infrastructure/v1alpha2"
+	"github.com/giantswarm/k8sclient/v3/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apiv1alpha2 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/cluster-operator/pkg/label"
+	"github.com/giantswarm/cluster-operator/pkg/project"
+	"github.com/giantswarm/cluster-operator/service/controller/key"
+)
+
+var (
+	clusterTransitionCreateDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystemCluster, "create_transition"),
+		"Latest cluster creation transition.",
+		[]string{
+			"cluster_id",
+			"release_version",
+		},
+		nil,
+	)
+	clusterTransitionUpdateDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystemCluster, "update_transition"),
+		"Latest cluster update transition.",
+		[]string{
+			"cluster_id",
+			"release_version",
+		},
+		nil,
+	)
+)
+
+type ClusterTransitionConfig struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+
+	NewCommonClusterObjectFunc func() infrastructurev1alpha2.CommonClusterObject
+}
+
+// ClusterTransition implements the ClusterTransition interface, exposing
+// cluster transition information.
+type ClusterTransition struct {
+	k8sClient                  k8sclient.Interface
+	logger                     micrologger.Logger
+	newCommonClusterObjectFunc func() infrastructurev1alpha2.CommonClusterObject
```

Comment:
Nitpick, either we separate the blocks consistently or we don't. I can see that such a callback can either be seen as dependency or setting. Let's just pick and stick to one. 

```suggestion

	newCommonClusterObjectFunc func() infrastructurev1alpha2.CommonClusterObject
```

Reply:
There are more places below. Let's talk if this is unclear.

Reply:
Oh sorry 🙈 , I was committing the suggestion without taking a second look if there's anything else which should be defined as dependency